queryparams/params.go

```diff
@@ -0,0 +1,239 @@
+package queryparams
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/ONSdigital/log.go/v2/log"
+)
+
+const (
+	Limit       = "limit"
+	Page        = "page"
+	Offset      = "offset"
+	SortName    = "sort"
+	DayBefore   = "before-day"
+	DayAfter    = "after-day"
+	MonthBefore = "before-month"
+	MonthAfter  = "after-month"
+	YearBefore  = "before-year"
+	YearAfter   = "after-year"
+	Keywords    = "keywords"
+	DateFrom    = "fromDate"
+	DateTo      = "toDate"
+	Published   = "type-published"
+	Cancelled   = "type-cancelled"
+	Upcoming    = "type-upcoming"
+	Provisional = "subtype-provisional"
+	Confirmed   = "subtype-confirmed"
+	Postponed   = "subtype-postponed"
+	Census      = "census"
+)
+
+func ParamGet(params url.Values, key string, defaultValue string) string {
+	valueAsString := params.Get(key)
+	if valueAsString == "" {
+		return defaultValue
+	}
+
+	return valueAsString
+}
+
+type IntValidator func(name string, valueAsString string) (int, error)
+
+func GetIntValidator(minValue, maxValue int) IntValidator {
+	return func(name string, valueAsString string) (int, error) {
+		value, err := strconv.Atoi(valueAsString)
+		if err != nil {
+			return 0, fmt.Errorf("%s search parameter provided with non numeric characters", name)
+		}
+		if value < minValue {
+			return 0, fmt.Errorf("%s search parameter provided with a value that is below the minimum value", name)
+		}
+		if value > maxValue {
+			return 0, fmt.Errorf("%s search parameter provided with a value that is above the maximum value", name)
+		}
+
+		return value, nil
+	}
+}
+
+var (
+	dayValidator   = GetIntValidator(1, 31)
+	monthValidator = GetIntValidator(1, 12)
+	yearValidator  = GetIntValidator(1900, 2150)
+)
+
```

Comment:
wondering if this code should be in api-clients ?

Reply:
Not sure if you mean this specific code fragment or the validators in general. I was thinking while doing this that the parameter parsing code is replicated pretty much everywhere there is a fronted, and I suppose I'm as guilty as anyone of doing it a particular way, but there is definitely a pattern that can be encoded in a library for parsing and validating string parameters

Reply:
I think I was concerned about the duplication of effort, validating these parameters in the release-calendar and then on the search endpoint. However I think the solution is for us to create our own validation library which is out of scope for this task

Reply:
yeah gotcha, the duplication of effort and code is a concern, and there is quite a bit going on in the code base at the mo'. And I agree with the validation lib as a solution (also it being out of scope of this task :-)